src/com/amazon/ion/impl/bin/IonManagedBinaryWriter.java

```diff
@@ -725,414 +116,276 @@ public boolean isFieldNameSet()
             return;
         }
-        token = intern(token);
-        user.writeSymbolToken(token);
+        currentWriter.writeSymbolToken(token);
     }
 
-    public void writeString(final String value) throws IOException
-    {
-        userState.writeString(this, value);
-        user.writeString(value);
+    public void writeString(final String value) throws IOException {
+        currentWriter.writeString(value);
     }
 
-    public void writeClob(byte[] data) throws IOException
+    @Override
+    public void writeString(byte[] data, int offset, int length) throws IOException
     {
-        user.writeClob(data);
+        user.writeString(data, offset, length);
     }
 
-    public void writeClob(final byte[] data, final int offset, final int length) throws IOException
-    {
-        user.writeClob(data, offset, length);
+    public void writeClob(byte[] data) throws IOException {
+        currentWriter.writeClob(data);
     }
 
-    public void writeBlob(byte[] data) throws IOException
-    {
-        user.writeBlob(data);
+    public void writeClob(final byte[] data, final int offset, final int length) throws IOException {
+        currentWriter.writeClob(data, offset, length);
     }
 
-    public void writeBlob(final byte[] data, final int offset, final int length) throws IOException
-    {
-        user.writeBlob(data, offset, length);
+    public void writeBlob(byte[] data) throws IOException {
+        currentWriter.writeBlob(data);
     }
 
-    @Override
-    public void writeString(byte[] data, int offset, int length) throws IOException
-    {
-        user.writeString(data, offset, length);
+    public void writeBlob(final byte[] data, final int offset, final int length) throws IOException {
+        currentWriter.writeBlob(data, offset, length);
     }
 
-    public void writeBytes(byte[] data, int off, int len) throws IOException
-    {
-        // this is a raw transfer--we basically have to dump the symbol table since we don't have much context
-        startLocalSymbolTableIfNeeded(/*writeIVM*/ true);
+    public void writeBytes(byte[] data, int off, int len) throws IOException {
         user.writeBytes(data, off, len);
     }
 
     // Stream Terminators
 
-    public void flush() throws IOException
-    {
-        if (getDepth() == 0 && localsLocked)
-        {
-            unsafeFlush();
-        }
-    }
-
-    private void unsafeFlush() throws IOException
-    {
-        if (user.hasWrittenValuesSinceFinished())
-        {
-            // this implies that we have a local symbol table of some sort and the user locked it
-            symbolState.closeTable(symbols);
+    public void flush() throws IOException {
+        if (getDepth() == 0) {
+            int maxId = lst.getMaxId();
+            if (!flushed && (user.hasWrittenValuesSinceFinished() || isStreamCopyOptimized()))
+                symbols.writeIonVersionMarker();
+            if (newSymbols || (!flushed && isStreamCopyOptimized()) || lstRequired) {
```

Comment:
I added this lstRequired boolean during the transition from the old external. I believe it will serialize an empty lst append on each flush if no new symbols have been written which isn't ideal.

Reply:
fixed it